File: BOJ/10711/10711.py
import sys
from collections import deque
input = sys.stdin.readline

height, width = map(int,input().split())
A = [list(map(lambda x: 0 if x == '.' else int(x),input().strip())) for _ in range(height)]
visited = [[False for x in range(width)] for y in range(height)]
dir = [[1,1],[1,0],[1,-1],[0,1],[0,-1],[-1,1],[-1,0],[-1,-1]]
ans = 0
q = deque()

# first Q
# 모래성과 인접한 물 부분으로 Q 초기화
for i in range(height):
    for j in range(width):
        if not A[i][j]:
            visited[i][j] = True
            for d in dir:
                ni, nj = i + d[0], j + d[1]
                if ni < 0 or nj < 0 or ni >= height or nj >= width:
                    continue
                if A[ni][nj]:
                    q.append([i,j])
                    break

while True:
    tempQ = deque()
    while len(q):
        x,y = q.popleft()
        for d in dir:
            nx, ny = x + d[0], y + d[1]
            if nx < 0 or ny < 0 or nx >= height or ny >= width or visited[nx][ny]:
                continue
            if A[nx][ny] > 0:
                visited[x][y] = True
                A[nx][ny] -= 1
                if not A[nx][ny]:
                    tempQ.append([nx,ny])
    if not len(tempQ):
        break;
    ans += 1
    q,tempQ = tempQ,q
    
print(ans)

File: BOJ/10711/10711.cpp
#include <cstdlib>
#include <iostream>
#include <queue>
using namespace std;
typedef pair<int,int> pii;

queue<pii> q;
int height, width;
int map[1001][1001] = {0,};
int visited[1001][1001] = {0,};
int dir[8][2] = { {1,1},{1,0},{1,-1},{0,1},{0,-1},{-1,1},{-1,0},{-1,-1} };

void get_input()
{
    cin >> height >> width;
    for (int i = 0; i < height; i++)
    {
        string s;
        cin >> s;
        for (int j = 0; j < width; j++) map[i][j] = s[j] == '.' ? 0 : s[j] - '0';
    }
}

void firstFill() // Q에 처음으로 삽입할 위치
{
    for (int i = 0; i < height; i++) 
    {
        for (int j = 0; j < width; j++) 
        {
            if (map[i][j]) continue;
            visited[i][j] = 1;
            for (int k = 0; k < 8; k++)
            {
                int ni = i + dir[k][0], nj = j + dir[k][1];
                if (ni < 0 || nj < 0 || ni >= height || nj >= width) continue;
                if (map[ni][nj])
                {
                    q.push({i,j});
                    break;
                }
            }
        }
    }
}

int solve()
{
    firstFill();
    int ret = 0;
    while (true)
    {
        queue<pii> tempQ;
        while (q.size())
        {
            int x = q.front().first, y = q.front().second;
            q.pop();
            for (int k = 0; k < 8; k++)
            {
                int nx = x + dir[k][0], ny = y + dir[k][1];
                if (nx < 0 || ny < 0 || nx >= height || ny >= width) continue;
                if (visited[nx][ny] || map[nx][ny] == 0) continue;
                visited[x][y] = true;
                if (map[nx][ny] > 0) map[nx][ny]--;
                if (map[nx][ny] == 0) tempQ.push({nx,ny});
            }
        }
        if (tempQ.size() == 0) break;
        ret++;
        swap(tempQ, q);
    }
    return ret;
}

int main()
{
    ios::sync_with_stdio(0);cin.tie(0);cout.tie(0);
    get_input();
    cout << solve() << '\n';
    return 0;
}